codec/gcm.go

```diff
@@ -101,6 +101,7 @@ func GCMEncoder(session *GCMSession, enc Encoder) Encoder {
 // GCMDEcoder accepts a GCMSession and a decoder that wraps data decryption
 func GCMDecoder(session *GCMSession, dec Decoder) Decoder {
 	return func(r io.Reader, buf []byte) (int, error) {
+		buf = buf[:len(buf)+16]
```

Comment:
Because we flipped the order of length prefix encoder/decoder and gcm encoder/decoder, length prefix now pre-pends and reads the length of the raw data instead of the encrypted data. Then gcm requires needs to add the amount of extra data expected in order to decode it

Reply:
That makes sense. Is the encrypted data length *always* 16 bytes longer than the raw data?

Reply:
Yes, it's always 16 bytes longer than the raw data. I have tested it with random data (both data and length). I don't see a problem with allocating a new buffer (since it should be a rare event), but a user will usually provide a buffer limited by their maximum message size. In which case allocating a new buffer can have a potential security issue where an attacker unloads a massive amount of data on you